File: zscala/src/main/scala/zingg/zs/Pairs.scala
package zingg.zs

import org.apache.spark.sql.DataFrame
import org.apache.spark.sql.functions.col

object PairBuilder {

  /** Candidate pairs from records sharing a block.
    *
    * `l_id < r_id` keeps each unordered pair once. When `crossSourceOnly` is
    * set (linkage mode), pairs from the same source are dropped so only
    * cross-source candidates survive — this is what makes N-way linkage fall
    * out of the same self-join: every source is canonicalised into one
    * DataFrame tagged with [[ZinggConf.SourceCol]], and the filter keeps only
    * pairs whose two records came from different systems.
    */
  def selfPairs(blocked: DataFrame, cfg: ZinggConf,
                blockCol: String = "z_block",
                crossSourceOnly: Boolean = false): DataFrame = {
    val left  = prefix(blocked, ZinggConf.LeftPrefix)
    val right = prefix(blocked, ZinggConf.RightPrefix)
    val lBlock = s"${ZinggConf.LeftPrefix}$blockCol"
    val rBlock = s"${ZinggConf.RightPrefix}$blockCol"
    val lId    = s"${ZinggConf.LeftPrefix}${cfg.idCol}"
    val rId    = s"${ZinggConf.RightPrefix}${cfg.idCol}"
    val joined = left.join(right, left(lBlock) === right(rBlock))
                     .filter(col(lId) < col(rId))
    if (crossSourceOnly) {
      val lSrc = s"${ZinggConf.LeftPrefix}${ZinggConf.SourceCol}"
      val rSrc = s"${ZinggConf.RightPrefix}${ZinggConf.SourceCol}"
      require(
        blocked.columns.contains(ZinggConf.SourceCol),
        s"cross-source pairing needs the '${ZinggConf.SourceCol}' column; canonicalise sources first"
      )
      joined.filter(col(lSrc) =!= col(rSrc))
    } else joined
  }

  private def prefix(df: DataFrame, p: String): DataFrame =
    df.columns.foldLeft(df) { (acc, c) => acc.withColumnRenamed(c, s"$p$c") }
}

File: python/zpy/src/zingg_py/pairs.py
"""Candidate pair generation — port of ``Pairs.scala`` (``PairBuilder``)."""

from __future__ import annotations

import pandas as pd

from . import blocking
from .blocking import BlockingTree, Leaf
from .schema import ZinggConf


def _prefix(df: pd.DataFrame, p: str) -> pd.DataFrame:
    return df.rename(columns={c: f"{p}{c}" for c in df.columns})


def self_pairs(
    blocked: pd.DataFrame,
    cfg: ZinggConf,
    block_col: str = blocking.BLOCK_COL,
    cross_source_only: bool = False,
) -> pd.DataFrame:
    """Candidate pairs from records sharing a block.

    ``l_id < r_id`` keeps each unordered pair once. A null block key never
    equi-joins, so signal-free records drop out instead of collapsing into one
    giant block — we filter null keys before the self-join (pandas would
    otherwise match ``None == None``).
    """
    # Drop rows with no blocking signal so they never pair.
    signal = blocked[blocked[block_col].notna()]
    left = _prefix(signal, ZinggConf.LEFT_PREFIX)
    right = _prefix(signal, ZinggConf.RIGHT_PREFIX)
    l_block = f"{ZinggConf.LEFT_PREFIX}{block_col}"
    r_block = f"{ZinggConf.RIGHT_PREFIX}{block_col}"
    l_id = f"{ZinggConf.LEFT_PREFIX}{cfg.id_col}"
    r_id = f"{ZinggConf.RIGHT_PREFIX}{cfg.id_col}"

    joined = left.merge(right, left_on=l_block, right_on=r_block)
    joined = joined[joined[l_id] < joined[r_id]]

    if cross_source_only:
        if ZinggConf.SOURCE_COL not in blocked.columns:
            raise ValueError(
                f"cross-source pairing needs the '{ZinggConf.SOURCE_COL}' column; "
                "canonicalise sources first"
            )
        l_src = f"{ZinggConf.LEFT_PREFIX}{ZinggConf.SOURCE_COL}"
        r_src = f"{ZinggConf.RIGHT_PREFIX}{ZinggConf.SOURCE_COL}"
        joined = joined[joined[l_src] != joined[r_src]]

    return joined.reset_index(drop=True)


def cold_start_pairs(
    df: pd.DataFrame,
    cfg: ZinggConf,
    cross_source_only: bool = False,
) -> pd.DataFrame:
    """Union of ``self_pairs`` over every cold-start canopy, deduped by id pair."""
    blockers = blocking.cold_start_blockers(cfg) or [Leaf("root")]
    per_blocker = [
        self_pairs(blocking.assign_blocks(df, b), cfg, cross_source_only=cross_source_only)
        for b in blockers
    ]
    l_id = f"{ZinggConf.LEFT_PREFIX}{cfg.id_col}"
    r_id = f"{ZinggConf.RIGHT_PREFIX}{cfg.id_col}"
    combined = pd.concat(per_blocker, ignore_index=True)
    if combined.empty:
        return combined
    return combined.drop_duplicates(subset=[l_id, r_id]).reset_index(drop=True)
